components/script/dom/paintworkletglobalscope.rs

```diff
@@ -54,44 +84,223 @@ impl PaintWorkletGlobalScope {
         }
     }
 
+    /// https://drafts.css-houdini.org/css-paint-api/#draw-a-paint-image
     fn draw_a_paint_image(&self,
                           name: Atom,
-                          concrete_object_size: Size2D<Au>,
-                          sender: Sender<Result<Image, PaintWorkletError>>) {
-        let width = concrete_object_size.width.to_px().abs() as u32;
-        let height = concrete_object_size.height.to_px().abs() as u32;
+                          size: Size2D<Au>,
+                          sender: Sender<Result<Image, PaintWorkletError>>)
+    {
+        // TODO: document paint definitions.
+        self.invoke_a_paint_callback(name, size, sender);
+    }
+
+    /// https://drafts.css-houdini.org/css-paint-api/#invoke-a-paint-callback
+    #[allow(unsafe_code)]
+    fn invoke_a_paint_callback(&self,
+                               name: Atom,
+                               size: Size2D<Au>,
+                               sender: Sender<Result<Image, PaintWorkletError>>)
+    {
+        let width = size.width.to_px().abs() as u32;
+        let height = size.height.to_px().abs() as u32;
+        debug!("Invoking a paint callback {}({},{}).", name, width, height);
+
+        let cx = self.worklet_global.get_cx();
+        let _ac = JSAutoCompartment::new(cx, self.worklet_global.reflector().get_jsobject().get());
+
+        // TODO: Steps 1-2.1.
+        // Step 3.
+        let definition = match self.paint_definitions.borrow().get(&name) {
+            Some(defn) => defn.clone(),
```

Comment:
This is scary when a PaintDefinition contains Heap values. We need to find something else to do here.